DIG-1355: Sonchau/UUID with fk

SonQBChau opened the pull request:
#### What's new:

- All tables will have FKs in the same manner as currently, but these FKs will be uuids
submitter_<object>_ids will be stored in each object as a text field.
- UUIDs will remain internal to katsu and not be exposed via any API.
- UUIDs will be generated by katsu when data is ingested
- The combination of (program_id, submitter_*_id) will have a UNIQUE constraint in the database (for donor, specimen, sample, treatment and follow-up ids)

#### Potentially Breaking changes:

- DonorWithClinicalData no longer has filters. Can only retrieve 1 specific object by providing the `program_id` and `submitter_*_id`.
- The structure of the DonorWithClinicalData request has changed and is now [`"/v2/authorized/donor_with_clinical_data/program/{program_id}/donor/{donor_id}"`](https://github.com/CanDIG/katsu/blob/2a7f44a9bf153215e4839427483c626aaee158a0/chord_metadata_service/mohpackets/docs/schema.json#L515)
- Ingest APIs no longer accept a list of object.
- Discovery query no longer has filters. This will be addressed in the next patch
- New data ingest script for development using 
`python chord_metadata_service/mohpackets/data/data_loader.py`

#### Package changes:

- **NEW** django-ninja for both API and schema + filter
- **NEW** orjson faster JSON package for renderer since we use a lot of json
- **REMOVE** drf, replaced by ninja, used serializer which is slower and hard to customize
- **REMOVE** django-filter, replaced by ninja, only work with drf
- **REMOVE** drf-spectacular, replaced by ninja, only work with drf. This used to generate schema based on the serializer to yml file. Since the new ninja only generate json, we need to convert it to yml

#### How to test:
- Due to significant changes, it is recommended to drop the database and re-create
- `pip install -r requirements/local.txt`
- `tox`

#### Related tickets:
[DIG-1389](https://candig.atlassian.net/browse/DIG-1389)
[DIG-1312](https://candig.atlassian.net/browse/DIG-1312)

[DIG-1389]: https://candig.atlassian.net/browse/DIG-1389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

SonQBChau committed:
WIP

SonQBChau committed:
Merge branch 'develop_candig' into sonchau/uuid_with_fk

SonQBChau committed:
Update apps.py

SonQBChau committed:
Update models.py

SonQBChau committed:
Delete pre_signals.py

SonQBChau committed:
Create signals.py

SonQBChau committed:
Update base.txt

SonQBChau committed:
Update models.py

SonQBChau committed:
Update signals.py

SonQBChau committed:
Update base.txt

SonQBChau committed:
Update utils.py

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update signals.py

SonQBChau committed:
Create schema.py

SonQBChau committed:
Create new_api.py

SonQBChau committed:
Update ingest.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update schema.py

filter schema WIP

SonQBChau committed:
Create api.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Update authentication.py

SonQBChau committed:
Update README.MD

SonQBChau committed:
Delete new_api.py

SonQBChau committed:
Update schema.py

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update utils.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update permissible_values.py

SonQBChau committed:
Update schema.py

SonQBChau committed:
Update utils.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Update permissible_values.py

SonQBChau committed:
Update schema.py

SonQBChau committed:
Delete locustfile.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Update ingest.py

SonQBChau committed:
Update models.py

SonQBChau committed:
Update schema.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Create 0003_alter_exposure_options_alter_donor_unique_together_and_more.py

SonQBChau committed:
Update models.py

SonQBChau committed:
Update schema.py

SonQBChau committed:
Update hooks.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Delete api.py

SonQBChau committed:
Create api.py

SonQBChau committed:
Create api_authorized.py

SonQBChau committed:
Create api_base.py

SonQBChau committed:
Create api_discovery.py

SonQBChau committed:
Create api_ingest.py

SonQBChau committed:
Delete api_authorized.py

SonQBChau committed:
Delete api_base.py

SonQBChau committed:
Delete api_discovery.py

SonQBChau committed:
Delete api_ingest.py

SonQBChau committed:
Delete schema.py

SonQBChau committed:
Create discovery_schema.py

SonQBChau committed:
Create donor_with_clinical_data_schema.py

SonQBChau committed:
Create filter_schema.py

SonQBChau committed:
Create ingest_schema.py

SonQBChau committed:
Create model_schema.py

SonQBChau committed:
Create schema.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update utils.py

SonQBChau committed:
refactor

SonQBChau committed:
Update schema.py

SonQBChau committed:
Update api_authorized.py

new CustomRouterPaginated

SonQBChau committed:
Update pagination.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Create 0004_alter_followup_submitter_primary_diagnosis_id_and_more.py

SonQBChau committed:
Update models.py

SonQBChau committed:
Update pagination.py

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update signals.py

SonQBChau committed:
Update test_models.py

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update test_models.py

fix specimenTest

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update test_models.py

SonQBChau committed:
Update test_models.py

fix treatment

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update test_models.py

fix chemo

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update test_models.py

fix immuno

SonQBChau committed:
Update test_models.py

fix radiation

SonQBChau committed:
Update test_models.py

SonQBChau committed:
Update model_schema.py

SonQBChau committed:
Update test_models.py

SonQBChau committed:
Update test_models.py

fix exposure

SonQBChau committed:
Update local.py

SonQBChau committed:
Update test_program.py

SonQBChau committed:
Update test_donor.py

SonQBChau committed:
Update factories.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update pagination.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update api.py

SonQBChau committed:
Update api_ingest.py

SonQBChau committed:
Update test_primary_diagnoses.py

SonQBChau committed:
Update factories.py

SonQBChau committed:
Update test_primary_diagnoses.py

SonQBChau committed:
Update test_sample_registration.py

SonQBChau committed:
Update test_specimen.py

SonQBChau committed:
Update test_treatment.py

SonQBChau committed:
Update api_authorized.py

SonQBChau committed:
Update test_chemotherapy.py

SonQBChau committed:
Update test_follow_up.py

SonQBChau committed:
Update test_treatment.py

SonQBChau committed:
Update api_discovery.py

SonQBChau committed:
Delete api.py

SonQBChau committed:
Delete api_authorized.py

SonQBChau committed:
Delete api_base.py

SonQBChau committed:
Delete api_discovery.py

SonQBChau committed:
Delete api_ingest.py

SonQBChau committed:
Create clinical_data.py

SonQBChau committed:
Create core.py

SonQBChau committed:
Create discovery.py

SonQBChau committed:
Create ingestion.py

SonQBChau committed:
Update apps.py

SonQBChau committed:
Delete authentication.py

SonQBChau committed:
Update ingest.py

SonQBChau committed:
Update README.MD

SonQBChau committed:
Update local.txt

SonQBChau committed:
Update dev.txt

SonQBChau committed:
Update base.txt

SonQBChau committed:
Update urls.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Delete hooks.py

SonQBChau committed:
Update utils.py

SonQBChau committed:
Delete urls.py

SonQBChau committed:
Delete throttling.py

SonQBChau committed:
Update test_models.py

SonQBChau committed:
Update test_chemotherapy.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update signals.py

SonQBChau committed:
Delete serializers_nested.py

SonQBChau committed:
Delete serializers.py

SonQBChau committed:
Create nested_data.py

SonQBChau committed:
Create model.py

SonQBChau committed:
Create ingestion.py

SonQBChau committed:
Create filter.py

SonQBChau committed:
Create discovery.py

SonQBChau committed:
Delete schema.py

SonQBChau committed:
Delete model_schema.py

SonQBChau committed:
Delete ingest_schema.py

SonQBChau committed:
Delete filter_schema.py

SonQBChau committed:
Delete donor_with_clinical_data_schema.py

SonQBChau committed:
Delete discovery_schema.py

SonQBChau committed:
Delete permissions.py

SonQBChau committed:
Update permissible_values.py

SonQBChau committed:
Update pagination.py

SonQBChau committed:
Update models.py

SonQBChau committed:
Delete filters.py

SonQBChau committed:
Update core.py

SonQBChau committed:
Update test_chemotherapy.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update core.py

SonQBChau committed:
Update utils.py

SonQBChau committed:
Update ingestion.py

SonQBChau committed:
Update ingest.py

SonQBChau committed:
Update core.py

SonQBChau committed:
Update clinical_data.py

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
Create schema.json

SonQBChau committed:
Update schema.json

SonQBChau committed:
Update model.py

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
Update clinical_data.py

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
make migrations

SonQBChau committed:
Update pagination.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update factories.py

SonQBChau committed:
fix linter

SonQBChau committed:
Update ingestion.py

SonQBChau committed:
Update discovery.py

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
Update ingestion.py

SonQBChau committed:
Update README.MD

SonQBChau committed:
Update core.py

SonQBChau committed:
Update base.py

SonQBChau committed:
Update filter.py

SonQBChau committed:
Update core.py

SonQBChau committed:
Merge branch 'develop_candig' into sonchau/uuid_with_fk

SonQBChau committed:
Update update-schema.yaml

SonQBChau committed:
Update core.py

SonQBChau committed:
Update schema.json

github-actions[bot] committed:
Update schema.json

SonQBChau committed:
Delete schema.yml

SonQBChau committed:
get donor with clinical data with full path

github-actions[bot] committed:
Update schema.json

github-actions[bot] committed:
Update openapi.md

SonQBChau committed:
rename ingest script

SonQBChau committed:
fix logging

SonQBChau committed:
clarify log

SonQBChau committed:
separate prod and dev settings

mshadbolt commented:
@SonQBChau are you able to add a line to the summary at the top that explains the change to the different package (including the package name) and how it might affect others - change from yaml to json schema, thank you!

SonQBChau commented:
I found a new bug with authorization in docker, temporary take it down

daisieh commented on .github/workflows/update-schema.yaml:
```diff
@@ -37,17 +41,17 @@ jobs:
       - name: Convert schema to OpenAPI documentation
         run: |
-          npx widdershins ./chord_metadata_service/mohpackets/docs/schema.yml -o ./chord_metadata_service/mohpackets/docs/openapi.md -u ./chord_metadata_service/mohpackets/docs/widdershins/templates/openapi3 -c true --omitHeader true
+          npx widdershins ./chord_metadata_service/mohpackets/docs/schema.json -o ./chord_metadata_service/mohpackets/docs/openapi.md -u ./chord_metadata_service/mohpackets/docs/widdershins/templates/openapi3 -c true --omitHeader true
```
Could you add a step to do something like https://stackoverflow.com/questions/38781929/how-to-convert-json-to-yaml-in-javascript, to minimize downstream changes?

SonQBChau commented:
> @SonQBChau are you able to add a line to the summary at the top that explains the change to the different package (including the package name) and how it might affect others - change from yaml to json schema, thank you!

I added the packages change

SonQBChau committed:
fix tqdm in prod env

Update schema.json

SonQBChau committed:
fix data_loader install permission

SonQBChau committed:
fix sampleregistration typo

SonQBChau committed:
fix typo in schema donor with clinical

daisieh reviewed (changes requested):
NestedFollowUpSchema does not have any properties in the schema.yml file.

There are still some typos being generated into the schema.yml/json files:
- primarydiagnosis instead of primary_diagnoses
- hormonetherapies instead of hormone_therapies
- samplesregistrations instead of sample_registrations
I think these are misnamed as properties of their parent NestedWhatever objects?

SonQBChau commented:
- NestedFollowUpSchema is just a typo, I put Schema instead of ModelSchema so it didn't show up.
- The schema is not being updated is a little complicated. This is a bug through github action. The new schema is generated, but it is not committed yet. However, the commit script can only read the old schema on github repo to get the SHA(which is a requirement to update), which is why the newly generated content is lost. I have to think about it.

github-actions[bot] committed:
Update schema.json

github-actions[bot] committed:
Update schema.yml

github-actions[bot] committed:
Update schema.md

github-actions[bot] committed:
Update schema.json

github-actions[bot] committed:
Update schema.json

github-actions[bot] committed:
Update schema.yml

github-actions[bot] committed:
Update schema.md

github-actions[bot] committed:
Update schema.yml

github-actions[bot] committed:
Update schema.md

daisieh reviewed (approved):
As far as I can tell, this works as expected. Looks good!

daisieh committed:
remove default value (makes fields optional)

https://stackoverflow.com/questions/67569529/using-fastapi-pydantic-how-do-i-define-an-optional-field-with-a-description

github-actions[bot] committed:
Update schema.json

github-actions[bot] committed:
Update schema.yml

github-actions[bot] committed:
Update schema.md

OrdiNeu reviewed (approved):
From the data-portal side everything looks fine as far as I can tell, will need some modifications to data portal code to get it working with this but nothing too out-of-the-ordinary :heavy_check_mark: .
